internal/monitors/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,276 @@
+			client.JVMStatsGroup:        conf.EnableEnhancedJVMStats,
+			client.ProcessStatsGroup:    conf.EnableEnhancedProcessStats,
+			client.ThreadpoolStatsGroup: conf.EnableEnhancedThreadPoolStats,
+			client.TransportStatsGroup:  conf.EnableEnhancedTransportStats,
+		})
+
+		// If current node is not the master, collect only node level stats
+		if !nodeIsCurrentMaster {
+			m.sendDatapoints(dps)
+			return
+		}
+
+		// Collect cluster level stats
+		clusterStatsOutput, err := esClient.GetClusterStats()
+
+		if err != nil {
+			logger.WithError(err).Errorf("Failed to GET cluster stats")
+			return
+		}
+
+		dps = append(dps, client.GetClusterStatsDatapoints(clusterStatsOutput, defaultDimensions, conf.EnableEnhancedClusterHealthStats)...)
+
+		if !*conf.EnableIndexStats {
+			m.sendDatapoints(dps)
+			return
+		}
+
+		// Collect Index stats
+		if !conf.IndexStatsMasterOnly || nodeIsCurrentMaster {
+			indexStatsOutput, err := esClient.GetIndexStats()
+
+			if err != nil {
+				logger.WithError(err).Errorf("Failed to GET index stats")
+				return
+			}
+
+			if conf.IndexSummaryOnly {
+				dps = append(dps, client.GetIndexStatsSummaryDatapoints(indexStatsOutput.AllIndexStats, defaultDimensions, utils.StringSliceToMap(conf.EnableEnhancedIndexStatsForIndexGroups), conf.EnableIndexStatsPrimaries)...)
+				m.sendDatapoints(dps)
+				return
+			}
+
+			dps = append(dps, client.GetIndexStatsDatapoints(indexStatsOutput.Indices, utils.StringSliceToMap(conf.Indexes), defaultDimensions, utils.StringSliceToMap(conf.EnableEnhancedIndexStatsForIndexGroups), conf.EnableIndexStatsPrimaries)...)
+		}
+
+		m.sendDatapoints(dps)
+	}, time.Duration(conf.IntervalSeconds)*time.Second)
+	return nil
+}
+
+// Prepares dimensions that are common to all datapoints from the monitor
+func prepareDefaultDimensions(userProvidedClusterName string, queriedClusterName *string) (map[string]string, error) {
+	dims := map[string]string{}
+	clusterName := userProvidedClusterName
+
+	if clusterName == "" {
+		if queriedClusterName == nil {
+			return nil, errors.New("Failed to GET cluster name from Elasticsearch API")
+		}
+		clusterName = *queriedClusterName
+	}
+
+	// "plugin_instance" dimension is added to maintain backwards compatibility with built-in content
+	dims["plugin_instance"] = clusterName
+	dims["cluster"] = clusterName
+	dims["plugin"] = monitorType
+
+	return dims, nil
+}
+
+func isCurrentMaster(nodeID string, masterInfoOutput *client.MasterInfoOutput) (bool, error) {
+	var isCurrentMaster bool
+
+	masterNode := masterInfoOutput.MasterNode
+
+	if masterNode == nil {
+		logger.Warn("Unable to identify Elasticsearch cluster master node. Assuming current node is not the current master.")
+		return isCurrentMaster, nil
+	}
+
+	if nodeID == *masterNode {
+		isCurrentMaster = true
+	}
+
+	return isCurrentMaster, nil
+}
+
+func prepareNodeMetricsDimensions(nodeStats map[string]client.NodeStats) (map[string]string, error) {
+	var nodeID *string
```

Comment:
This really doesn't need to be a pointer.  You can just check if it is `== ""` below if you need to check if it didn't come through.  It is safer than dealing with pointer deferencing just in case you missed something, especially when refactoring, that could cause a panic.

Reply:
Made it a string